### ISSUE TITLE
Rename base types

### ISSUE THREAD

**Issue by uncomputable** [BlockstreamResearch/simfony#19]
## TL;DR

"Base type" is a generic name that might be confusing.

## Details

The Simfony type system comes with "fundamental" types: unit `()`, sum `Either<A, B>` and product `(A, B)`.

Other Simfony types are defined as "macros": `Option<A>` = `Either<(), A>`, `bool` = `Either<(), ()>`, ...

At the moment, the Simfony type system expands all macros and works at the "fundamental" type level.

We want to pick a name for "fundamental" types that distinguishes them from "macro" types. Currently the former are called "base types".

Alternatively, we could enforce explicit casting between types. On the surface, this would make all types disjoint. Types with equal structure (`Option<A>` ↔ `Either<(), A>` etc.) would be available for casting.

**Comment by apoelstra** [comment]
We should use the term "structural types" which [is a standard term for types which are defined (solely) by their structure](https://en.wikipedia.org/wiki/Structural_type_system).

That wiki page suggests the term "nominative types" for types which have names/explicit declarations, which I hadn't heard before but which I like.

**Comment by uncomputable** [comment]
Simfony's type system has since been overhauled. There are no more base types.
